design/cdk-bootstrap.md

```diff
@@ -0,0 +1,199 @@
+        }
+      ]
+    }
+  }
+  ```
+
+* An IAM role, called the **Action role**,
+  that will be assumed when executing the CloudFormation deployment actions
+  (CreateChangeSet and ExecuteChangeSet) in the continuous delivery pipeline.
+  It is also assumable by any principal from the account(s) passed by the `trust` option.
+  
+  This role has the following permissions:
+  
+  ```json
+  {
+    "PolicyDocument": {
+      "Statement": [
+        {
+          "Action": [
+            "cloudformation:CreateChangeSet",   "cloudformation:DeleteChangeSet",
+            "cloudformation:DescribeChangeSet", "cloudformation:DescribeStacks",
+            "cloudformation:ExecuteChangeSet"
+          ],
+          "Resource": "*"
+        },
+        {
+          "Action": "iam:PassRole",
+          "Resource": "arn-of-the-cfn-deployment-role"
+        }
+      ]
+    }
+  }
+  ```
+
+* An IAM role, called the **CFN deployment role**,
+  that will be used to perform the actual CFN stack deployment in the continuous delivery pipeline to this environment.
+  It is assumable *only* by the CloudFormation service principal
+  (this is for security reasons, as this role will have, necessarily, very wide permissions).
+  It will not have any inline policies,
+  but will instead have the Managed Policies attached that the user selected when prompted (see above).
+
+#### Physical resource names
+
+All of the above resources will be created with well-defined physical names -
+none of them will rely on automated CloudFormation naming.
+This is so that the other stages of CDK synthesis,
+like asset resolution, can rely on the concrete names
+(there is no reliable way to transfer this kind of information in an automated way across regions and/or accounts).
+This also allows for more fine-grained permissions -
+for instance, the continuous delivery pipeline needs to grant the **action role**
+permissions to read from the pipeline's bucket
+(to read the artifact that contains the CFN template to deploy);
+this way, it can add a well-defined name to the resource policy of the artifact bucket,
+instead of granting those permissions to all principals in the account.
+
+The naming scheme will include the following elements in order to minimize the chance of name collisions:
+
+* The region we're bootstrapping in.
+* The account ID we're bootstrapping in.
+* The type of the resource (file assets bucket, Docker assets repository, etc.).
+* An optional *qualifier* that defends against S3 bucket sniping.
+
+### CLI options in detail
+
+These options are inherited from the current CLI experience:
+
+* `--profile`: use the given local AWS credentials profile when interacting with the target environment.
+
+* `--stack-name`, `-s`: allows you to explicitly name the CloudFormation bootstrap stack
+  (instead of relying on the default naming scheme).
+
+* `--bucket-name`, `-b`: allows you to explicitly name the file assets S3 bucket
```

Comment:
I think we should just let users pass in their own template if they want to control this (e.g. what about ECR repo name, IAM role names, etc).

Reply:
Parameter for each of those I guess...? I think a a custom template is a little much - what guarantees we have that the template will have all of the requires resources, and named correctly..?

Reply:
Let's just allow passing in a template for such customisations. Remember that if users change the names of these resources, they will have to also override the `addXxxAsset` methods of `core.Stack` to return their custom names. It's not sufficient to just change the names here. So I rather we just leave this as a general customisation option as oppose to specific degrees of freedom. This also aligns with the tenet that the bootstrap template should be a static fixed file, not generated.

Reply:
For now, we have a bunch of parameters that control the names of the S3 bucket and ECR repo. I hope this is enough for V1.

Reply:
We don’t need any control in V1. As I commented above, if people want to customize the names they can just use their own template.

Reply:
Like we discussed, let's:

1. Remove these customizations.
2. Explain that any customizations can be done by deploying a custom bootstrap template.
3. Have an option (`cdk bootstrap --show-template`?) that allows displaying (and saving to a file with `cdk bootstrap --show-template > template.json`) the template the tool uses.